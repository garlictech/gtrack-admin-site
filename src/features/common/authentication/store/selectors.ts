import { Injectable } from '@angular/core';

import _get from 'lodash-es/get';

import { createFeatureSelector, createSelector, MemoizedSelector } from '@ngrx/store';

import { DeepstreamSelectors } from '@bit/garlictech.angular-features.common.deepstream-ngx/src/store/selectors';

import { User } from '../interfaces';
import { Auth, AuthenticationState, featureName, JwtAuthState } from './state';

const selectAuthFeature = createFeatureSelector<AuthenticationState>(featureName);

const selectJwtFeature = createSelector(
  selectAuthFeature,
  (state: AuthenticationState) => state.jwtAuth
);

const selectAuth = createSelector(
  selectJwtFeature,
  (state: JwtAuthState) => state.auth
);

export const selectUser = createSelector(
  selectAuth,
  (state: Auth) => _get(state, 'user')
);

export const selectRole = createSelector(
  selectUser,
  (user: User) => _get(user, 'roles[0]')
);
export const selectUserId = createSelector(
  selectUser,
  (user: User) => _get(user, 'id')
);
export const loggedOut = createSelector(
  selectUser,
  user => !user
);

@Injectable({
  providedIn: 'root'
})
export class AuthenticationSelectors {
  auth: MemoizedSelector<object, Auth>;
  user: MemoizedSelector<object, User>;
  loggingIn: MemoizedSelector<object, boolean>;
  jwtLoggingIn: MemoizedSelector<object, boolean>;
  loggedIn: MemoizedSelector<object, boolean>;
  loggedOut: MemoizedSelector<object, boolean>;
  magicLinkEmailSent: MemoizedSelector<object, boolean>;
  loginFailed: MemoizedSelector<object, boolean>;
  role;
  selectedRole;
  token: MemoizedSelector<object, string>;
  termsAccepted: MemoizedSelector<object, boolean>;
  loginRefused: MemoizedSelector<object, boolean>;
  selectUserId: MemoizedSelector<object, string>;

  constructor(private readonly _deepstreamSelectors: DeepstreamSelectors) {
    const selectUiFeature = createSelector(
      selectAuthFeature,
      (state: AuthenticationState) => state.uiState
    );

    this.auth = selectAuth;
    this.user = selectUser;

    this.jwtLoggingIn = createSelector(
      selectJwtFeature,
      jwtState => jwtState.loggingIn
    );

    this.loggingIn = createSelector(
      this.jwtLoggingIn,
      this._deepstreamSelectors.loggingIn,
      (jwtLoggingIn, dsState) => jwtLoggingIn || dsState
    );

    this.loggedOut = loggedOut;

    this.token = createSelector(
      selectAuth,
      state => _get(state, 'token')
    );

    this.loggedIn = createSelector(
      this.token,
      this._deepstreamSelectors.loggedIn,
      (token, dsState) => !!token && dsState
    );

    this.magicLinkEmailSent = createSelector(
      selectJwtFeature,
      state => state.emailSent
    );

    this.loginFailed = createSelector(
      selectJwtFeature,
      this._deepstreamSelectors.failed,
      (jwtState, deepstreamFailed) => jwtState.failed || !!deepstreamFailed
    );

    this.role = selectRole;
    this.selectUserId = selectUserId;

    this.selectedRole = createSelector(
      selectUiFeature,
      state => state.selectedRole
    );

    this.termsAccepted = createSelector(
      selectUiFeature,
      state => state.termsAccepted
    );

    this.loginRefused = createSelector(
      selectUiFeature,
      state => state.loginRefused
    );
  }
}
