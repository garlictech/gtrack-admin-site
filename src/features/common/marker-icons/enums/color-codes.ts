// color names: https://www.colorhexa.com/
export enum EColorCodes {
  BLUE = '#3975D7',
  BROWN = '#9D7051',
  CYAN = '#5EC8BD',
  DARK_GREEN = '#118E4D',
  DARK_MAGENTA = '#74286B',
  GREY = '#A8A8A8',
  LIGHT_GREEN = '#66C547',
  ORANGE = '#FF8A22',
  PURPLE = '#8C4EB8',
  RED = '#F34648',
  // Custom
  HIGHLIGHTED = '#FF2905',
  // For admin
  PIN = '#2299E5'
}
