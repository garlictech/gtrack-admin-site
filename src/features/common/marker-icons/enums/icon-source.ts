export enum EIconSource {
  ICON = 'icon',
  MARKER = 'marker'
}
