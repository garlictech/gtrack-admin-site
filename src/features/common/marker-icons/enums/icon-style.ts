export enum EIconStyle {
  DEFAULT = 'default',
  HIGHLIGHTED = 'highlighted',
  PIN = 'pin',
  // for checkpoints
  GREEN = 'green',
  GREY = 'grey',
  RED = 'red'
}
