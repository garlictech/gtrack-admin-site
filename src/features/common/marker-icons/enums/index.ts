export * from './color-codes';
export * from './default-color-map';
export * from './icon-source';
export * from './icon-style';
