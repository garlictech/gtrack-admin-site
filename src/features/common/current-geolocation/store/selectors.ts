import { createFeatureSelector, createSelector } from '@ngrx/store';
import { CurrentGeolocationState, featureName } from './state';

export const selectFeature = createFeatureSelector<CurrentGeolocationState>(featureName);

export const selectTracking = createSelector(
  selectFeature,
  (state: CurrentGeolocationState) => state.tracking
);

export const selectCurrentLocation = createSelector(
  selectFeature,
  (state: CurrentGeolocationState) => state.currentLocation
);
