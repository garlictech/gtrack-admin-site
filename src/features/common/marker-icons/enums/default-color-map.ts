import { EColorCodes } from './color-codes';

export const DEFAULT_COLOR_MAP = {
  asterisco: EColorCodes.LIGHT_GREEN,
  atm: EColorCodes.BLUE,
  bank: EColorCodes.BLUE,
  bicycle_parking: EColorCodes.BROWN,
  bicycle: EColorCodes.CYAN,
  bus_station: EColorCodes.BROWN,
  cafe: EColorCodes.PURPLE,
  cave_entrance: EColorCodes.LIGHT_GREEN,
  checkpoint: EColorCodes.DARK_MAGENTA,
  city: EColorCodes.LIGHT_GREEN,
  cliff: EColorCodes.DARK_GREEN,
  drinking_water: EColorCodes.DARK_GREEN,
  finish: EColorCodes.ORANGE,
  fuel: EColorCodes.BROWN,
  hiking: EColorCodes.ORANGE,
  hunting_stand: EColorCodes.ORANGE,
  lodging: EColorCodes.PURPLE,
  marketplace: EColorCodes.DARK_GREEN,
  mud: EColorCodes.DARK_GREEN,
  parking: EColorCodes.BROWN,
  peak: EColorCodes.DARK_GREEN,
  pharmacy: EColorCodes.RED,
  photo: EColorCodes.GREY,
  picnic: EColorCodes.LIGHT_GREEN,
  place_of_worship: EColorCodes.LIGHT_GREEN,
  police: EColorCodes.BLUE,
  post_box: EColorCodes.BLUE,
  post_office: EColorCodes.BLUE,
  pub: EColorCodes.PURPLE,
  recycling: EColorCodes.DARK_GREEN,
  restaurant: EColorCodes.PURPLE,
  shelter: EColorCodes.LIGHT_GREEN,
  sight: EColorCodes.LIGHT_GREEN,
  spring: EColorCodes.DARK_GREEN,
  start: EColorCodes.ORANGE,
  start_finish: EColorCodes.ORANGE,
  store: EColorCodes.CYAN,
  sunsetland: EColorCodes.DARK_GREEN,
  telephone: EColorCodes.BLUE,
  toilets: EColorCodes.RED,
  train_station: EColorCodes.BROWN,
  tree: EColorCodes.DARK_GREEN,
  waste_basket: EColorCodes.BLUE,
  water: EColorCodes.DARK_GREEN
};
