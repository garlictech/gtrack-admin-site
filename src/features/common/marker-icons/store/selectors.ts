import { createFeatureSelector, createSelector } from '@ngrx/store';

import { EIconStyle } from '../enums';
import { getSvgContent } from '../pure';
import { svgCircleContentAdapter, svgIconContentAdapter, svgMarkerContentAdapter } from './reducer';
import { featureName, State, SvgContent } from './state';

export const featureSelector = createFeatureSelector<State>(featureName);

const svgIconSelector = createSelector(
  featureSelector,
  (state: State) => state.icons
);
export const getAllSvgIcons = svgIconContentAdapter.getSelectors(svgIconSelector).selectAll;
export const getAllSvgIconsCount = svgIconContentAdapter.getSelectors(svgIconSelector).selectTotal;

export const getIcon = (type: string, encoded: boolean, iconStyle: EIconStyle) =>
  createSelector(
    getAllSvgIcons,
    (svgContents: Array<SvgContent>) => getSvgContent(type, svgContents, encoded, iconStyle)
  );

const svgMarkerSelector = createSelector(
  featureSelector,
  (state: State) => state.markers
);
export const getAllSvgMarkers = svgMarkerContentAdapter.getSelectors(svgMarkerSelector).selectAll;
export const getAllSvgMarkersCount = svgMarkerContentAdapter.getSelectors(svgMarkerSelector).selectTotal;

export const getMarker = (type: string, encoded: boolean, iconStyle: EIconStyle) =>
  createSelector(
    getAllSvgMarkers,
    (svgContents: Array<SvgContent>) => getSvgContent(type, svgContents, encoded, iconStyle)
  );

const svgCircleSelector = createSelector(
  featureSelector,
  (state: State) => state.circles
);
export const getAllSvgCircles = svgCircleContentAdapter.getSelectors(svgCircleSelector).selectAll;
export const getAllSvgCirclesCount = svgCircleContentAdapter.getSelectors(svgCircleSelector).selectTotal;

export const getCircle = (type: string, encoded: boolean, iconStyle: EIconStyle) =>
  createSelector(
    getAllSvgCircles,
    (svgContents: Array<SvgContent>) => getSvgContent(type, svgContents, encoded, iconStyle)
  );
