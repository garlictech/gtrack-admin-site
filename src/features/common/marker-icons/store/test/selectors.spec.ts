import { EIconStyle } from '../../enums';
import * as fromSelectors from '../selectors';
import { createState, createSvgContent, createSvgContentEntityState } from './utils';

const _state = createState({
  markers: createSvgContentEntityState([createSvgContent('atm'), createSvgContent('bank'), createSvgContent('cafe')]),
  icons: createSvgContentEntityState([createSvgContent('atm'), createSvgContent('bank')])
});

const testCases = [
  {
    name: 'getAllSvgIcons',
    selector: fromSelectors.getAllSvgIcons,
    state: { ..._state }
  },
  {
    name: 'getAllSvgIconsCount',
    selector: fromSelectors.getAllSvgIconsCount,
    state: { ..._state }
  },
  {
    name: 'getAllSvgMarkers',
    selector: fromSelectors.getAllSvgMarkers,
    state: { ..._state }
  },
  {
    name: 'getAllSvgMarkersCount',
    selector: fromSelectors.getAllSvgMarkersCount,
    state: { ..._state }
  },
  {
    name: 'getIcon - encoded - default',
    selector: fromSelectors.getIcon('atm', true, EIconStyle.DEFAULT),
    state: { ..._state }
  },
  {
    name: 'getMarker - encoded - default',
    selector: fromSelectors.getMarker('atm', true, EIconStyle.DEFAULT),
    state: { ..._state }
  },
  {
    name: 'getIcon - not encoded - default',
    selector: fromSelectors.getIcon('atm', false, EIconStyle.DEFAULT),
    state: { ..._state }
  },
  {
    name: 'getMarker - not encoded - default',
    selector: fromSelectors.getMarker('atm', false, EIconStyle.DEFAULT),
    state: { ..._state }
  },
  {
    name: 'getIcon - not encoded - highlighted',
    selector: fromSelectors.getIcon('atm', false, EIconStyle.HIGHLIGHTED),
    state: { ..._state }
  },
  {
    name: 'getMarker - not encoded - highlighted',
    selector: fromSelectors.getMarker('atm', false, EIconStyle.HIGHLIGHTED),
    state: { ..._state }
  }
];

testCases.forEach(({ name, selector, state }) => {
  test(`${name} with input ${JSON.stringify(state)}`, () => {
    expect(selector(state)).toMatchSnapshot();
  });
});
