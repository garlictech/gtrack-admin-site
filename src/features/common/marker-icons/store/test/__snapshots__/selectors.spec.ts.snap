// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllSvgIcons with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `
Array [
  Object {
    "content": "<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>",
    "id": "atm",
  },
  Object {
    "content": "<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>",
    "id": "bank",
  },
]
`;

exports[`getAllSvgIconsCount with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `2`;

exports[`getAllSvgMarkers with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `
Array [
  Object {
    "content": "<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>",
    "id": "atm",
  },
  Object {
    "content": "<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>",
    "id": "bank",
  },
  Object {
    "content": "<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>",
    "id": "cafe",
  },
]
`;

exports[`getAllSvgMarkersCount with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `3`;

exports[`getIcon - encoded - default with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `"data:image/svg+xml;base64,PHN2Zz48cGF0aCBpZD0iaWNvbiIgZmlsbD0iIzM5NzVENyI+PC9wYXRoPjwvc3ZnPg=="`;

exports[`getIcon - not encoded - default with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `"<svg><path id=\\"icon\\" fill=\\"#3975D7\\"></path></svg>"`;

exports[`getIcon - not encoded - highlighted with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `"<svg><path id=\\"icon\\" fill=\\"#FF2905\\"></path></svg>"`;

exports[`getMarker - encoded - default with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `"data:image/svg+xml;base64,PHN2Zz48cGF0aCBpZD0iaWNvbiIgZmlsbD0iIzM5NzVENyI+PC9wYXRoPjwvc3ZnPg=="`;

exports[`getMarker - not encoded - default with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `"<svg><path id=\\"icon\\" fill=\\"#3975D7\\"></path></svg>"`;

exports[`getMarker - not encoded - highlighted with input {"features.common.marker-icons":{"markers":{"ids":["atm","bank","cafe"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"cafe":{"id":"cafe","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}},"icons":{"ids":["atm","bank"],"entities":{"atm":{"id":"atm","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"},"bank":{"id":"bank","content":"<svg><path id=\\"icon\\" fill=\\"#7C7CF0\\"></path></svg>"}}}}} 1`] = `"<svg><path id=\\"icon\\" fill=\\"#FF2905\\"></path></svg>"`;
